### Task-2 Model Devlopment 

####  Befoe each Model i did SMOTE because the class is highly immbalanced. 
## Data Preprocessing

### Class Imbalance
Initially, the dataset exhibited significant class imbalance, with the distribution of classes as follows:

| Class       | Count    |
|-------------|----------|
| Fraud       | 12268      |
| Not Fraud   | 116878  |

This imbalance posed challenges for model training, as machine learning algorithms may tend to favor the majority class.

### Applying SMOTE
To address this issue, we applied SMOTE for only training datasets, which generates synthetic examples of the minority class (fraud) to create a more balanced dataset. After applying SMOTE, the new class distribution was as follows:

| Class       | Count    |
|-------------|----------|
| Fraud       | 187004  |
| Not Fraud   | 187004  |

With the application of SMOTE, the dataset now contains an equal number of samples for both classes, allowing for a more effective model training process.


I applied six different models to the balanced dataset, logging key metrics to monitor performance. The models included:

1. **Logistic Regression**
2. **Decision Tree Classifier**
3. **Random Forest Classifier**
4. **Gradient Boosting Classifier**
5. **Multi-Layer Perceptron (MLP)**
6. **LSTM and RNN Models and RNN** 

In [ ]:
import sys 
import pandas as pd
sys.path.insert(0, '/10 A KAI 2/WEEK 8&9/Fraud_detection/')

In [2]:
from script.model1 import  FraudModel

In [3]:
Data = pd.read_csv('../Data/merged.csv', index_col=0)

In [17]:
Data['class'].value_counts()

class
0    116878
1     12268
Name: count, dtype: int64

In [4]:
Data = Data.drop(columns=['lower_bound_ip_address','upper_bound_ip_address','time_diff', 'transaction_count','signup_time','purchase_time' ,'user_transaction_frequency','device_transaction_frequency', 'user_transaction_velocity','device_transaction_velocity','device_id','ip_address', 'device_id_encoded'], axis=1)

In [5]:
model = FraudModel(Data, dataset_name="Fraud", source="Fraud Data Source")

In [6]:
model.fraud_preprocess()

purchase_value       age  class  purchase_day_of_week  signup_hour  \
31545         0.931750 -0.943490      0              0.989675    -1.229144   
97542         0.222055  0.565460      0              0.989675     0.652288   
12873         0.549607 -0.943490      0             -0.006798    -0.650242   
104500       -0.105497 -1.639928      0             -0.505034    -0.794967   
21011        -1.524887 -0.130978      0             -0.006798     0.218111   
...                ...       ...    ...                   ...          ...   
54852         2.460324  3.002994      0             -0.505034     0.797013   
96464        -0.924375  0.449387      0             -0.006798    -1.229144   
129697       -0.214680  0.217241      0              0.491439    -0.939693   
120623        1.586853  1.726191      0             -0.505034    -0.360791   
42662         1.150118 -0.943490      1             -1.501506     1.665366   

        signup_day  signup_month  purchase_day  purchase_month  \
31545    -0.425555     -1.390901      0.649434       -1.500438   
97542    -1.456606      0.822662      1.217198        1.125064   
12873     1.521987      0.379949     -0.258989        0.749993   
104500   -0.540116     -1.390901     -1.280965       -1.125366   
21011    -1.342045     -0.948188     -0.372542       -1.125366   
...            ...           ...           ...             ...   
54852    -1.227484      0.822662      1.671409        1.125064   
96464     1.751109      1.265374     -1.621624        1.500136   
129697   -0.540116     -0.505476     -1.167412       -0.000151   
120623    1.751109      0.822662     -0.940307        1.500136   
42662     0.147251     -0.505476      0.535881       -0.750295   

        purchase__hour  ...  category_Direct  category_SEO  category_Chrome  \
31545        -0.218131  ...                0             1                1   
97542         1.373715  ...                1             0                0   
12873        -0.073418  ...                0             1                1   
104500        1.229002  ...                0             0                0   
21011         0.071295  ...                1             0                0   
...                ...  ...              ...           ...              ...   
54852        -0.652271  ...                1             0                0   
96464        -0.652271  ...                1             0                0   
129697        1.663142  ...                1             0                0   
120623        1.229002  ...                1             0                1   
42662        -0.796984  ...                0             0                1   

        category_FireFox  category_IE  category_Opera  category_Safari  \
31545                  0            0               0                0   
97542                  0            0               0                1   
12873                  0            0               0                0   
104500                 0            0               0                1   
21011                  0            1               0                0   
...                  ...          ...             ...              ...   
54852                  0            0               0                1   
96464                  0            1               0                0   
129697                 1            0               0                0   
120623                 0            0               0                0   
42662                  0            0               0                0   

        category_F  category_M  Country_encoded  
31545            1           0         0.020932  
97542            1           0        -0.328521  
12873            0           1         0.020932  
104500           0           1         1.505499  
21011            0           1         0.020932  
...            ...         ...              ...  
54852            0           1        -0.011600  
96464            1           0        -1

In [7]:
import mlflow
mlflow.start_run()

<ActiveRun: >

In [8]:
model.model_logistic_regression()

2024/10/22 22:42:03 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/22 22:42:03 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/22 22:42:03 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/22 22:42:12 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/22 22:42:18 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "d:\10 A KAI 2\WEEK 8&9\Fraud_detection\venv\Lib\site-packages\mlflow\types\utils.py:407: UserWarning: Hint: Inferred schema contains integer column(s). In

Accuracy: 0.84
Logging metrics for model: Logistic Regression
Dataset name: Fraud, Source: Fraud Data Source


2024/10/22 22:42:29 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


LogisticRegression()

In [9]:
model.model_decision_tree()

2024/10/22 22:42:30 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/22 22:42:30 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/22 22:42:30 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/22 22:42:33 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/22 22:42:38 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "d:\10 A KAI 2\WEEK 8&9\Fraud_detection\venv\Lib\site-packages\mlflow\types\utils.py:407: UserWarning: Hint: Inferred schema contains integer column(s). In

Accuracy: 0.90
Logging metrics for model: Decision Tree
Dataset name: Fraud, Source: Fraud Data Source


2024/10/22 22:42:49 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


DecisionTreeClassifier()

In [11]:
model.model_random_forest()

2024/10/22 22:43:18 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/22 22:43:18 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/22 22:43:18 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/22 22:43:23 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/22 22:43:28 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "d:\10 A KAI 2\WEEK 8&9\Fraud_detection\venv\Lib\site-packages\mlflow\types\utils.py:407: UserWarning: Hint: Inferred schema contains integer column(s). In

Accuracy: 0.96


2024/10/22 22:47:43 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


RandomForestClassifier(random_state=42)

In [12]:
model.model_gradient_boosting()

2024/10/22 22:47:47 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/22 22:47:47 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/22 22:47:47 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/22 22:48:19 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/22 22:48:24 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "d:\10 A KAI 2\WEEK 8&9\Fraud_detection\venv\Lib\site-packages\mlflow\types\utils.py:407: UserWarning: Hint: Inferred schema contains integer column(s). In

Accuracy: 0.95
Logging metrics for model: Gradient Boosting
Dataset name: Fraud, Source: Fraud Data Source


2024/10/22 22:49:43 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


GradientBoostingClassifier()

In [13]:
# Other model calls...
model.model_mlp()

2024/10/22 22:49:44 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/22 22:49:44 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/22 22:49:44 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/22 22:49:48 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


Epoch 1/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 14s 2ms/step - accuracy: 0.8030 - loss: 0.4207
Epoch 2/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 20s 2ms/step - accuracy: 0.8721 - loss: 0.3063
Epoch 3/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 21s 2ms/step - accuracy: 0.8721 - loss: 0.2996
Epoch 4/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 10s 2ms/step - accuracy: 0.8724 - loss: 0.2960
Epoch 5/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 10s 2ms/step - accuracy: 0.8724 - loss: 0.2914
Epoch 6/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 10s 2ms/step - accuracy: 0.8743 - loss: 0.2874
Epoch 7/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 11s 2ms/step - accuracy: 0.8740 - loss: 0.2849
Epoch 8/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 19s 2ms/step - accuracy: 0.8758 - loss: 0.2804
Epoch 9/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 10s 2ms/step - accuracy: 0.8779 - loss: 0.2757
Epoch 10/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 10s 2ms/step - accuracy: 0.8786 - loss: 0.2751
808/808 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step


2024/10/22 22:52:12 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!


Accuracy: 0.93
Logging metrics for model: MLP
Dataset name: Fraud, Source: Fraud Data Source


2024/10/22 22:52:15 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/22 22:52:15 WARNING mlflow.keras.save: You are saving a Keras model without specifying model signature.
2024/10/22 22:53:02 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


5844/5844 ━━━━━━━━━━━━━━━━━━━━ 10s 2ms/step
An error occurred: All fields in the model signature's input schema must be of type TensorSpec.


d:\10 A KAI 2\WEEK 8&9\Fraud_detection\venv\Lib\site-packages\mlflow\types\utils.py:407: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(


<Sequential name=sequential, built=True>

In [14]:
model.model_lstm()

2024/10/22 22:53:25 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/22 22:53:25 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/22 22:53:25 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/22 22:53:44 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
d:\10 A KAI 2\WEEK 8&9\Fraud_detection\venv\Lib\site-packages\keras\src\layers\rnn\rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer 

Epoch 1/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 18s 2ms/step - accuracy: 0.8139 - loss: 0.4073
Epoch 2/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 13s 2ms/step - accuracy: 0.8715 - loss: 0.3091
Epoch 3/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 13s 2ms/step - accuracy: 0.8683 - loss: 0.3076
Epoch 4/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 23s 3ms/step - accuracy: 0.8716 - loss: 0.2977
Epoch 5/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 14s 2ms/step - accuracy: 0.8730 - loss: 0.2929
Epoch 6/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 14s 2ms/step - accuracy: 0.8750 - loss: 0.2862
Epoch 7/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 17s 3ms/step - accuracy: 0.8750 - loss: 0.2843
Epoch 8/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 13s 2ms/step - accuracy: 0.8770 - loss: 0.2781
Epoch 9/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 14s 2ms/step - accuracy: 0.8797 - loss: 0.2743
Epoch 10/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 13s 2ms/step - accuracy: 0.8794 - loss: 0.2741
808/808 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step


2024/10/22 22:56:19 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!


Accuracy: 0.94
Logging metrics for model: LSTM
Dataset name: Fraud, Source: Fraud Data Source


2024/10/22 22:56:22 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/22 22:56:22 WARNING mlflow.keras.save: You are saving a Keras model without specifying model signature.
2024/10/22 22:56:41 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/22 22:56:41 WARNING mlflow.keras.save: You are saving a Keras model without specifying model signature.
2024/10/22 22:56:49 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


<Sequential name=sequential_1, built=True>

In [15]:
model.model_rnn()

2024/10/22 22:56:50 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/22 22:56:50 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/22 22:56:50 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/22 22:56:54 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


Epoch 1/10


d:\10 A KAI 2\WEEK 8&9\Fraud_detection\venv\Lib\site-packages\keras\src\layers\rnn\rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


5844/5844 ━━━━━━━━━━━━━━━━━━━━ 14s 2ms/step - accuracy: 0.8004 - loss: 0.4213
Epoch 2/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 14s 2ms/step - accuracy: 0.8734 - loss: 0.3041
Epoch 3/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 13s 2ms/step - accuracy: 0.8717 - loss: 0.2987
Epoch 4/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 12s 2ms/step - accuracy: 0.8741 - loss: 0.2927
Epoch 5/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 12s 2ms/step - accuracy: 0.8738 - loss: 0.2887
Epoch 6/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 12s 2ms/step - accuracy: 0.8759 - loss: 0.2821
Epoch 7/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 11s 2ms/step - accuracy: 0.8753 - loss: 0.2810
Epoch 8/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 10s 2ms/step - accuracy: 0.8791 - loss: 0.2742
Epoch 9/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 11s 2ms/step - accuracy: 0.8782 - loss: 0.2745
Epoch 10/10
5844/5844 ━━━━━━━━━━━━━━━━━━━━ 11s 2ms/step - accuracy: 0.8791 - loss: 0.2707
808/808 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step


2024/10/22 22:58:58 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!


Accuracy: 0.91
Logging metrics for model: RNN
Dataset name: Fraud, Source: Fraud Data Source


2024/10/22 22:59:01 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/22 22:59:02 WARNING mlflow.keras.save: You are saving a Keras model without specifying model signature.
2024/10/22 22:59:10 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/22 22:59:10 WARNING mlflow.keras.save: You are saving a Keras model without specifying model signature.
2024/10/22 22:59:19 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


<Sequential name=sequential_2, built=True>

In [16]:
model.display_metrics_table()

Model  Accuracy  Precision    Recall  F1 Score       AUC
0  Logistic Regression  0.838482   0.319234  0.618174  0.421038  0.739892
1        Decision Tree  0.898258   0.471400  0.584352  0.521834  0.757782
2    Gradient Boosting  0.954201   0.942241  0.551752  0.695965  0.774101
3                  MLP  0.928378   0.640074  0.562347  0.598698  0.764575
4                 LSTM  0.944444   0.796393  0.557865  0.656123  0.771446
5                  RNN  0.912350   0.536149  0.574165  0.554506  0.761009

### Performance Summary:

1. **Gradient Boosting** achieved the highest accuracy (95.42%) and precision (94.22%), indicating that it correctly identified a high proportion of positive cases, though it had a lower recall (55.18%), suggesting it missed some positive cases.

2. **Logistic Regression** had the lowest precision (31.92%), indicating many of its positive predictions were false positives, but a moderate recall (61.82%).

3. **LSTM** and **MLP** models performed well with good accuracy (94.44% and 92.84%, respectively) and decent F1 scores, indicating a balanced performance between precision and recall.

4. **Decision Tree** provided a good balance with an accuracy of 89.83%, but both precision and recall were lower compared to other models, suggesting room for improvement in correctly identifying true positives.

5. **RNN** had an overall solid performance but did not stand out in any metric compared to the others.

Overall, **Gradient Boosting** showed the best performance in accuracy and precision, while the **LSTM** model had a strong overall balance in various metrics.

## I will work to improve each model performance!!